Detectors/PHOS/reconstruction/src/CaloRawFitterGS.cxx

```diff
@@ -0,0 +1,317 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file CaloRawFitterGS.cxx
+/// \author Dmitri Peresunko
+
+#include <gsl/span>
+
+#include "PHOSReconstruction/CaloRawFitterGS.h"
+#include "PHOSBase/PHOSSimParams.h"
+
+using namespace o2::phos;
+CaloRawFitterGS::CaloRawFitterGS() : CaloRawFitter()
+{
+  mDecTime = o2::phos::PHOSSimParams::Instance().mSampleDecayTime;
+  mTimeAccuracy = o2::phos::PHOSSimParams::Instance().mSampleTimeFitAccuracy;
+  mAmpAccuracy = o2::phos::PHOSSimParams::Instance().mSampleAmpFitAccuracy;
+  init();
+}
+
+void CaloRawFitterGS::init()
+{
+  //prepare fitting arrays, once per lifetime
+  double k = o2::phos::PHOSSimParams::Instance().mSampleDecayTime;
+  ma0[0] = 1.;
+  mb0[0] = 1.;
+  mb1[0] = 0.;
+  mb2[0] = 0.;
+  mb3[0] = 0.;
+  mb4[0] = 0.;
+  for (int i = 1; i < NMAXSAMPLES; i++) {
+    double xi = k * i;
+    ma0[i] = exp(-xi);
+    mb0[i] = mb0[i - 1] + ma0[i];
+    mb1[i] = mb1[i - 1] + ma0[i] * xi;
+    mb2[i] = mb2[i - 1] + ma0[i] * xi * xi;
+    mb3[i] = mb3[i - 1] + ma0[i] * xi * xi * xi;
+    mb4[i] = mb4[i - 1] + ma0[i] * xi * xi * xi * xi;
+  }
+}
+
+CaloRawFitterGS::FitStatus CaloRawFitterGS::evaluate(gsl::span<short unsigned int> signal)
+{
+
+  //Pedestal analysis mode
+  if (mPedestalRun) {
+    int nPed = signal.size();
+    mAmp = 0.;
+    mTime = 0.;
+    for (auto a : signal) {
+      mAmp += a;
+      mTime += a * a;
+    }
+    if (nPed > 0) {
+      mAmp /= nPed;
+      mTime = mTime / nPed - mAmp * mAmp;
+      if (mTime > 0.) {
+        mTime = sqrt(mTime);
+      }
+    }
+    mOverflow = false;
+    mStatus = kOK;
+    return kOK;
+  }
+
+  mStatus = kNotEvaluated;
+  // Extract amplitude and time
+  mStatus = evalFit(signal);
+  return mStatus;
+}
+
+CaloRawFitterGS::FitStatus CaloRawFitterGS::evalFit(gsl::span<short unsigned int> signal)
+{
```

Comment:
here too, I guess this method is used a lot, you can save quite some multiplications by precalculating some terms or using Horner form for polynomials